File: src/test/rustdoc-js/doc-alias.js
// exact-check

const QUERY = [
    'StructItem',
    'StructFieldItem',
    'StructMethodItem',
    'ImplTraitItem',
    'StructImplConstItem',
    'ImplTraitFunction',
    'EnumItem',
    'VariantItem',
    'EnumMethodItem',
    'TypedefItem',
    'TraitItem',
    'TraitTypeItem',
    'AssociatedConstItem',
    'TraitFunctionItem',
    'FunctionItem',
    'ModuleItem',
    'ConstItem',
    'StaticItem',
    'UnionItem',
    'UnionFieldItem',
    'UnionMethodItem',
    'MacroItem',
];

const EXPECTED = [
    {
        // StructItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Struct',
                'alias': 'structitem',
                'href': '../doc_alias/struct.Struct.html',
                'is_alias': true
            },
        ],
    },
    {
        // StructFieldItem
        'others': [
            {
                'path': 'doc_alias::Struct',
                'name': 'field',
                'alias': 'structfielditem',
                'href': '../doc_alias/struct.Struct.html#structfield.field',
                'is_alias': true
            },
        ],
    },
    {
        // StructMethodItem
        'others': [
            {
                'path': 'doc_alias::Struct',
                'name': 'method',
                'alias': 'structmethoditem',
                'href': '../doc_alias/struct.Struct.html#method.method',
                'is_alias': true
            },
        ],
    },
    {
        // ImplTraitItem
        'others': [],
    },
    {
        // StructImplConstItem
        'others': [
            {
                'path': 'doc_alias::Struct',
                'name': 'ImplConstItem',
                'alias': 'structimplconstitem',
                'href': '../doc_alias/struct.Struct.html#associatedconstant.ImplConstItem',
                'is_alias': true
            },
        ],
    },
    {
        // ImplTraitFunction
        'others': [
            {
                'path': 'doc_alias::Struct',
                'name': 'function',
                'alias': 'impltraitfunction',
                'href': '../doc_alias/struct.Struct.html#method.function',
                'is_alias': true
            },
        ],
    },
    {
        // EnumItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Enum',
                'alias': 'enumitem',
                'href': '../doc_alias/enum.Enum.html',
                'is_alias': true
            },
        ],
    },
    {
        // VariantItem
        'others': [
            {
                'path': 'doc_alias::Enum',
                'name': 'Variant',
                'alias': 'variantitem',
                'href': '../doc_alias/enum.Enum.html#variant.Variant',
                'is_alias': true
            },
        ],
    },
    {
        // EnumMethodItem
        'others': [
            {
                'path': 'doc_alias::Enum',
                'name': 'method',
                'alias': 'enummethoditem',
                'href': '../doc_alias/enum.Enum.html#method.method',
                'is_alias': true
            },
        ],
    },
    {
        // TypedefItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Typedef',
                'alias': 'typedefitem',
                'href': '../doc_alias/type.Typedef.html',
                'is_alias': true
            },
        ],
    },
    {
        // TraitItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Trait',
                'alias': 'traititem',
                'href': '../doc_alias/trait.Trait.html',
                'is_alias': true
            },
        ],
    },
    {
        // TraitTypeItem
        'others': [
            {
                'path': 'doc_alias::Trait',
                'name': 'Target',
                'alias': 'traittypeitem',
                'href': '../doc_alias/trait.Trait.html#associatedtype.Target',
                'is_alias': true
            },
        ],
    },
    {
        // AssociatedConstItem
        'others': [
            {
                'path': 'doc_alias::Trait',
                'name': 'AssociatedConst',
                'alias': 'associatedconstitem',
                'href': '../doc_alias/trait.Trait.html#associatedconstant.AssociatedConst',
                'is_alias': true
            },
        ],
    },
    {
        // TraitFunctionItem
        'others': [
            {
                'path': 'doc_alias::Trait',
                'name': 'function',
                'alias': 'traitfunctionitem',
                'href': '../doc_alias/trait.Trait.html#tymethod.function',
                'is_alias': true
            },
        ],
    },
    {
        // FunctionItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'function',
                'alias': 'functionitem',
                'href': '../doc_alias/fn.function.html',
                'is_alias': true
            },
        ],
    },
    {
        // ModuleItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Module',
                'alias': 'moduleitem',
                'href': '../doc_alias/Module/index.html',
                'is_alias': true
            },
        ],
    },
    {
        // ConstItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Const',
                'alias': 'constitem',
                'href': '../doc_alias/constant.Const.html',
                'is_alias': true
            },
            {
                'path': 'doc_alias::Struct',
                'name': 'ImplConstItem',
            },
        ],
    },
    {
        // StaticItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Static',
                'alias': 'staticitem',
                'href': '../doc_alias/static.Static.html',
                'is_alias': true
            },
        ],
    },
    {
        // UnionItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Union',
                'alias': 'unionitem',
                'href': '../doc_alias/union.Union.html',
                'is_alias': true
            },
            // Not an alias!
            {
                'path': 'doc_alias::Union',
                'name': 'union_item',
                'href': '../doc_alias/union.Union.html#structfield.union_item'
            },
        ],
    },
    {
        // UnionFieldItem
        'others': [
            {
                'path': 'doc_alias::Union',
                'name': 'union_item',
                'alias': 'unionfielditem',
                'href': '../doc_alias/union.Union.html#structfield.union_item',
                'is_alias': true
            },
        ],
    },
    {
        // UnionMethodItem
        'others': [
            {
                'path': 'doc_alias::Union',
                'name': 'method',
                'alias': 'unionmethoditem',
                'href': '../doc_alias/union.Union.html#method.method',
                'is_alias': true
            },
        ],
    },
    {
        // MacroItem
        'others': [
            {
                'path': 'doc_alias',
                'name': 'Macro',
                'alias': 'macroitem',
                'href': '../doc_alias/macro.Macro.html',
                'is_alias': true
            },
        ],
    },
];

File: tests/rustdoc-js/doc-alias.rs
/// Doc for `Struct`
#[doc(alias = "StructItem")]
pub struct Struct {
    /// Doc for `Struct`'s `field`
    #[doc(alias = "StructFieldItem")]
    pub field: u32,
}

impl Struct {
    /// Doc for `Struct::ImplConstItem`
    #[doc(alias = "StructImplConstItem")]
    pub const ImplConstItem: i32 = 0;
    /// Doc for `Struct::method`
    #[doc(alias = "StructMethodItem")]
    pub fn method(&self) {}
}

impl Trait for Struct {
    type Target = u32;
    const AssociatedConst: i32 = 12;

    /// Doc for `Trait::function` implemented for Struct
    #[doc(alias = "ImplTraitFunction")]
    fn function() -> Self::Target {
        0
    }
}

/// Doc for `Enum`
#[doc(alias = "EnumItem")]
pub enum Enum {
    /// Doc for `Enum::Variant`
    #[doc(alias = "VariantItem")]
    Variant,
}

impl Enum {
    /// Doc for `Enum::method`
    #[doc(alias = "EnumMethodItem")]
    pub fn method(&self) {}
}

/// Doc for type alias `Typedef`
#[doc(alias = "TypedefItem")]
pub type Typedef = i32;

/// Doc for `Trait`
#[doc(alias = "TraitItem")]
pub trait Trait {
    /// Doc for `Trait::Target`
    #[doc(alias = "TraitTypeItem")]
    type Target;
    /// Doc for `Trait::AssociatedConst`
    #[doc(alias = "AssociatedConstItem")]
    const AssociatedConst: i32;

    /// Doc for `Trait::function`
    #[doc(alias = "TraitFunctionItem")]
    fn function() -> Self::Target;
}

/// Doc for `function`
#[doc(alias = "FunctionItem")]
pub fn function() {}

/// Doc for `Module`
#[doc(alias = "ModuleItem")]
pub mod Module {}

/// Doc for `Const`
#[doc(alias = "ConstItem")]
pub const Const: u32 = 0;

/// Doc for `Static`
#[doc(alias = "StaticItem")]
pub static Static: u32 = 0;

/// Doc for `Union`
#[doc(alias = "UnionItem")]
pub union Union {
    /// Doc for `Union::union_item`
    #[doc(alias = "UnionFieldItem")]
    pub union_item: u32,
    pub y: f32,
}

impl Union {
    /// Doc for `Union::method`
    #[doc(alias = "UnionMethodItem")]
    pub fn method(&self) {}
}

/// Doc for `Macro`
#[doc(alias = "MacroItem")]
#[macro_export]
macro_rules! Macro {
    () => {};
}
